pkg/daemon/drain.go

```diff
@@ -0,0 +1,113 @@
+package daemon
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubectl/pkg/drain"
+)
+
+func (dn *Daemon) drainRequired() bool {
+	// Drain operation is not useful on a single node cluster as there
+	// is no other node in the cluster where workload with PDB set
+	// can be rescheduled. It can lead to node being stuck at drain indefinitely.
+	// These clusters can take advantage of graceful node shutdown feature.
+	// https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown
+	return !isSingleNodeTopology(dn.getControlPlaneTopology())
+}
+
+func (dn *Daemon) cordonOrUncordonNode(desired bool) error {
+	backoff := wait.Backoff{
+		Steps:    5,
+		Duration: 10 * time.Second,
+		Factor:   2,
+	}
+	var lastErr error
+	if err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		err := drain.RunCordonOrUncordon(dn.drainer, dn.node, desired)
+		if err != nil {
+			lastErr = err
+			glog.Infof("cordon/uncordon failed with: %v, retrying", err)
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		if err == wait.ErrWaitTimeout {
+			return errors.Wrapf(lastErr, "failed to cordon/uncordon node (%d tries): %v", backoff.Steps, err)
+		}
+		return errors.Wrap(err, "failed to cordon/uncordon node")
+	}
+	return nil
+}
+
+func (dn *Daemon) drain() error {
+	backoff := wait.Backoff{
+		Steps:    5,
+		Duration: 10 * time.Second,
+		Factor:   2,
+	}
+	var lastErr error
+	if err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		err := drain.RunNodeDrain(dn.drainer, dn.node.Name)
+		if err != nil {
+			lastErr = err
+			glog.Infof("Draining failed with: %v, retrying", err)
+			return false, nil
+		}
+		return true, nil
+
+	}); err != nil {
+		if err == wait.ErrWaitTimeout {
+			failMsg := fmt.Sprintf("%d tries: %v", backoff.Steps, lastErr)
+			MCDDrainErr.WithLabelValues(dn.node.Name, "WaitTimeout").Set(float64(backoff.Steps))
+			dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeWarning, "FailedToDrain", failMsg)
+			return errors.Wrapf(lastErr, "failed to drain node (%d tries): %v", backoff.Steps, err)
+		}
+		MCDDrainErr.WithLabelValues(dn.node.Name, "UnknownError").Set(float64(backoff.Steps))
+		dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeWarning, "FailedToDrain", err.Error())
+		return errors.Wrap(err, "failed to drain node")
+	}
+
+	return nil
+}
+
+func (dn *Daemon) performDrain() error {
+	// Skip drain process when we're not cluster driven
+	if dn.kubeClient == nil {
```

Comment:
I know this is from our old code, but just curious, do you know if we would ever hit this condition?

Reply:
yeah, we do hit it when update() gets called during firstboot as firstboot is not cluster driven.

My initial goal was to move this logic into drainRequired() but it was complicating some logic because during firstboot dn.node is null and it causes to crash things.